timur/lib/client/jsx/components/browser/browser.jsx

```diff
@@ -67,21 +77,62 @@ export default function Browser({model_name, record_name, tab_name}) {
   );
   const {view, record, tab, revision, template, can_edit} = browserState;
   const [mode, setMode] = useState('loading');
-  const loading = !view || !template || !record || !tab_name;
+  const [error, setError] = useState(null);
   const {cancelEdits, approveEdits} = useEditActions(setMode, browserState);
-  const {selectOrShowTab, selectDefaultTab, selectTab, showTab} = useTabActions(
-    browserState,
-    setMode
-  );
+  const requestDocuments = useRequestDocuments();
+
+  const browseToTab = useCallback((tabName) => {
+    invoke(setLocation(window.location.href.replace(/#.*/, '') + `#${tabName}`));
+  }, [invoke, setLocation]);
 
   // Set at 'skin' on the browser styling.
   let skin = 'browser';
   if (mode === 'browse') skin = 'browser ' + model_name;
   const editMode = useCallback(() => setMode('edit'), [setMode]);
 
-  // On mount
+  // const loading = !view || !template || !record || !tab_name;
+  const [loading, loadDocuments] = useAsyncWork(function* loadDocuments(model_name, view) {
+    setMode('loading');
+
+    if (!template) yield invoke(requestModel(model_name));
+    if (!view) {
+      // we are told the model and record name, get the view
+      view = yield invoke(requestView(model_name));
+    }
+
+    if (!tab_name) {
+      browseToTab(getDefaultTab(view));
+      return;
+    }
+
+    const tab = view.tabs.find(t=>t.name == tab_name)
+
+    // set tab
+    if (!tab) return Promise.reject('Could not find tab by the name ' + tab);
+    let exchange_name = `tab ${tab.name} for ${model_name} ${record_name}`;
+    let attribute_names = getAttributes(tab);
+
+    let hasAttributes =
+      record &&
+      template &&
+      attribute_names.every(attr_name => attr_name in record);
+
+    // ensure attribute data is present in the document
+    if (!hasAttributes) {
+      // or else make a new request
+      yield requestDocuments({
+        model_name,
+        record_names: [record_name],
+        attribute_names,
+        exchange_name,
+      });
+    }
+
+    setMode('browse');
+  }, { cancelWhenChange: [] });
+
+  // On mount, check if we have model and record names, and if not, navigate away and reload this component.
   useEffect(() => {
```

Comment:
This useEffect is now a one time on mount hook.  It either redirects the browser or starts loading data.  Once.